components/calendar/src/ethiopic.rs

```diff
@@ -0,0 +1,204 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+//! This module contains types and implementations for the Ethiopic calendar
+
+use crate::coptic::Coptic;
+use crate::iso::{Iso, IsoYear};
+use crate::julian::Julian;
+use crate::{
+    types, ArithmeticDate, Calendar, CalendarArithmetic, Date, DateDuration, DateDurationUnit,
+    DateTime, DateTimeError,
+};
+use core::marker::PhantomData;
+use tinystr::tinystr;
+
+#[derive(Copy, Clone, Debug, Hash, Default, Eq, PartialEq)]
+// The Ethiopic Calendar
+pub struct Ethiopic;
+
+#[derive(Copy, Clone, Debug, Hash, Eq, PartialEq)]
+pub struct EthiopicDateInner(ArithmeticDate<Ethiopic>);
+
+impl CalendarArithmetic for Ethiopic {
+    fn month_days(year: i32, month: u8) -> u8 {
+        if (1..=12).contains(&month) {
+            30
+        } else if month == 13 {
+            if Self::is_leap_year(year) {
+                6
+            } else {
+                5
+            }
+        } else {
+            0
+        }
+    }
+
+    fn months_for_every_year() -> u8 {
+        13
+    }
+
+    fn is_leap_year(year: i32) -> bool {
+        year % 4 == 3
+    }
+}
+
+impl Calendar for Ethiopic {
+    type DateInner = EthiopicDateInner;
+    fn date_from_iso(&self, iso: Date<Iso>) -> EthiopicDateInner {
+        let fixed_iso = Iso::fixed_from_iso(*iso.inner());
+        Self::ethiopic_from_fixed(fixed_iso)
+    }
+
+    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
+        let fixed_ethiopic = Ethiopic::fixed_from_ethiopic(date.0);
+        Iso::iso_from_fixed(fixed_ethiopic)
+    }
+
+    fn months_in_year(&self, date: &Self::DateInner) -> u8 {
+        date.0.months_in_year()
+    }
+
+    fn days_in_year(&self, date: &Self::DateInner) -> u32 {
+        date.0.days_in_year()
+    }
+
+    fn days_in_month(&self, date: &Self::DateInner) -> u8 {
+        date.0.days_in_month()
+    }
+
+    fn day_of_week(&self, date: &Self::DateInner) -> types::IsoWeekday {
+        Iso.day_of_week(Ethiopic.date_to_iso(date).inner())
+    }
+
+    fn offset_date(&self, date: &mut Self::DateInner, offset: DateDuration<Self>) {
+        date.0.offset_date(offset);
+    }
+
+    #[allow(clippy::field_reassign_with_default)]
+    fn until(
+        &self,
+        date1: &Self::DateInner,
+        date2: &Self::DateInner,
+        _largest_unit: DateDurationUnit,
+        _smallest_unit: DateDurationUnit,
+    ) -> DateDuration<Self> {
+        date1.0.until(date2.0, _largest_unit, _smallest_unit)
+    }
+
+    fn year(&self, date: &Self::DateInner) -> types::Year {
+        crate::gregorian::year_as_gregorian(date.0.year)
```

Comment:
https://docs.google.com/document/d/1LIL4oXfiDK9X7Zd2eHanMxlLEqjHEUHBC0eS4PKvwlw/edit#heading=h.7hirz9yhgnbl

Reply:
I discussed this a bit with @sffc , I think the plan going forward will be that we only support `ethiopic` as far as formatting identifiers are concerned, _however_ each EthiopicDateInner contains an `alem` bool, which, when `true`, disables the second era.

On further inspection it does seem like the Alem (single-era) and Mihret (two era) dates do have three separate era names: Alem has the single Anno Mundi era, whereas Mihret has a negative "before incarnation" and positive "after incarnation" pair of eras. ICU's handling of this is buggy.

We'd need to update the CLDR code to.

For this PR, we can use era code names `"incarnation"` (counts upwards from 8 CE), `"before-incar"` (counts downwards from 8 CE), and `"mundi"` (counts upwards from 5493 BC) or something, using the bool to switch between incarnation and mundi representations.

Reply:
Getting the data merged on the CLDR transform side will be a little annoying but is doable.

Reply:
I've filed https://unicode-org.atlassian.net/browse/ICU-21985 

For now, I think you should match ICU's behavior (there are two eras, both count forwards) and we can fix it up once there's a good answer there.